### PR TITLE
[WIP] sage: 8.9 -> 9.2

### PR DESCRIPTION
**Pull request by omasanori**
###### Motivation for this change

This change updates SageMath to the latest version, migrates to Python 3, and gets rid of most of the current workarounds.

Depends on:
- <del>https://github.com/NixOS/nixpkgs/pull/101327</del>
- <del>https://github.com/NixOS/nixpkgs/pull/101382</del>
- <del>https://github.com/NixOS/nixpkgs/pull/101830</del>
- <del>https://github.com/NixOS/nixpkgs/pull/101876</del>

Part of https://github.com/NixOS/nixpkgs/issues/101964

TODO:
- Make `env-locations`, `sage-env` and `sage-with-env` suitable with the new upstream scripts.
  - `env-locations`: patch `src/bin/sage-env-config.in` and `build/pkgs/sage_conf/src/sage_conf.py` instead (?)
  - `sage-env`: use upstream `sage-env` instead (?)
  - `sage-with-env`: TBD
- Run tests and look at failures.
- Update documentation if needed.
- Check `.nix` files if more cleanup and/or simplification is possible.

This is work-in-progress, but any comments or suggestions are welcome!

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/science/math/sage/sagelib.nix [hunk 1]
```diff
@@ -51,6 +52,8 @@
 , libbraiding
 , gmpy2
 , pplpy
+, sqlite # FIXME
```
Comment (SuperSandro2000) [inline pkgs/applications/science/math/sage/sagelib.nix:55]: What needs to be fixed here?

Comment (omasanori) [reply pkgs/applications/science/math/sage/sagelib.nix:55]: To determine where the dependencies come from. The dependency to `gaic` is explicitly mentioned in [the official document](https://wiki.sagemath.org/ReleaseTours/sage-9.2):

> The packages giacpy_sage and sage_brial have been merged into sagelib as sage.libs.giac and sage.rings.polynomial.pbori.

However, I have no idea about SQLite and Boehm GC (libgc). libgc especially *smells;* there might be a lack of build-time configuration or something. That is why I marked here.

Surely they are not clear to others, I will add some more words. Thank you for your comment!

**Comment by yurrriq** [comment]
I'm having trouble getting this to build locally, but here are a couple things I noticed so far.

- `/nix/store/fydrv699rcrsk8h35n79z2w8cm0iv2lx-sage-with-env-9.2.rc3/bin/sage: line 532: /nix/store/fydrv699rcrsk8h35n79z2w8cm0iv2lx-sage-with-env-9.2.rc3/bin/python3: No such file or directory` -- I guess we need to run `patchShebangs`
- remove `sagenb` from `sagedoc.buildInputs` (in `pkgs/applications/science/math/sage/sagedoc.nix`)

**Comment by omasanori** [comment]
@yurrriq Thank you for your report! It helps me much.

**Comment by SuperSandro2000** [comment]
@omasanori Please rebase the PR on master to fix the eval django_2_2 borg eval error.

**Comment by omasanori** [comment]
@SuperSandro2000 Done. Thanks!

**Comment by omasanori** [comment]
Breaking news: [SageMath 9.2 was just released!](https://github.com/sagemath/sage/releases/tag/9.2) 🎉

**Comment by SuperSandro2000** [comment]
@omasanori Can you rebase again to fix ``undefined variable 'boolToString'``? Sorry about the hassle.

**Comment by omasanori** [comment]
No problem.

**Comment by jcumming** [comment]
@omasanori ; this is awesome! Does the testsuite pass?

**Comment by omasanori** [comment]
@jcumming Thank you for your interest!
No, it just compiles Cython code yet. I have been trying to adopt the Nix-specific scripts to newer upstream framework. Running the test suites is definitely the next step and I hope I will reach there soon.

**Comment by yurrriq** [comment]
I've got the day off tomorrow, anything in particular I can help with?

**Comment by omasanori** [comment]
@yurrriq Sorry, no significant output from me this week. I hope I will take some time next week...

Anyway, if your machine is capable of some build tasks, could you review recent PRs (Python or math libraries especially)? It would be really helpful, as it will reduce the possibility of failures in dependencies which sometimes queued on this branch.
Even doing non-SageMath-related-ones matters, as it will decrease core members' duties. Ones with a "1 - 10" or "10 - 100" label should be relatively safe for your spare (CPU-)time.

**Comment by refnil** [comment]
Hi @omasanori 
I've been trying to make some progress on this merge request today. First, I had to rebase on a commit after 734c7444071243ba157fef71a880ed1b072abf11 since it fix a hanging problem in `pymzq` test.
Then, I had to put back the line `echo '# do nothing' >  src/bin/sage-env-config` to make sage compile at all. I saw you removed it in one of your commit. Is it because you intend the patch all the places that could call this file? Anyway, it was simpler to just create the file like before.

Then, I had to add link to python3 inside the `sage-with-env` derivation since sage now expect `$SAGE_LOCAL/bin/python3` to exist. Again, I don't know if we'd patch sage or add a link to python there. 
You can see the line [there](https://github.com/refnil/nixpkgs/blob/d6b8bca5f808e4636805ff09becaacd6b0096c0b/pkgs/applications/science/math/sage/sage-with-env.nix#L133) on my repo.

After that, it looked like `sagelib` did all of its cytonization. I could not get further however. Sage itself and the tests would't run.
I've attached the log of trying to run the test. The line `ModuleNotFoundError: No module named 'Cython'` made me think that it could be a problem with `cython` dynamic linking. I tried poking around adding `cython` dependencies here and there but it did not prove useful yet.
[sage-problem.txt](https://github.com/NixOS/nixpkgs/files/5569274/sage-problem.txt)

You can find my branch [there](https://github.com/refnil/nixpkgs/pull/3).

**Comment by omasanori** [comment]
@refnil Thank you for sharing your experiments! Your results are in fact informative.

> Then, I had to put back the line echo '# do nothing' > src/bin/sage-env-config to make sage compile at all. I saw you removed it in one of your commit. Is it because you intend the patch all the places that could call this file? Anyway, it was simpler to just create the file like before.

Sure, my intention was to use upstream's tooling so that we don't have to implement our own implementation of a similar mechanism but I am fine with leaving it anyway.

> Then, I had to add link to python3 inside the sage-with-env derivation since sage now expect $SAGE_LOCAL/bin/python3 to exist. Again, I don't know if we'd patch sage or add a link to python there.

I will check this point, as, patching the correct places would be better if possible.

**Comment by refnil** [comment]
While looking at the failing test stack trace, I saw that the test is calling libpynac which in turn is calling libpython2.7. For one, the appearance of python 2.7 seem suspicious. As you can see at the stack frame #4, the signal handler is called directly after the libpython2.7 call. I'll try to make pynac work with python 3.8.

```
#4  <signal handler called>
No symbol table info available.
#5  0x00007fff9cd7d4ab in PyInt_FromLong ()
   from /nix/store/xdha1qdqlh9hhymxq3w79zvjmwdyxqyz-python-2.7.18/lib/libpython2.7.so.1.0
No symbol table info available.
#6  0x00007fff9cfb4c87 in _GLOBAL__sub_I_numeric.cpp ()
   from /nix/store/jrdwhgc3dcrj8l9mxwh474ikkax89rh8-pynac-0.7.26/lib/libpynac.so.18
No symbol table info available.
#7  0x00007ffff7fe19ee in call_init ()
   from /nix/store/5didcr1sjp2rlx8abbzv92rgahsarqd9-glibc-2.32/lib/ld-linux-x86-64.so.2
No symbol table info available.
```

**Comment by refnil** [comment]
Moving pynac to python3 fixed to crash mentioned above. I tried opening sage repl and a jupyter notebook and both worked. However, it looked like some tests failed and one in particular hanged. Do you think we should add the commit to this PR or add a new one for pynac? You can find the code [there](https://github.com/refnil/nixpkgs/commit/30afa507a3acf39910fcf278008e71c6a1e3ebb2).

**Comment by omasanori** [comment]
Excellent! Would you make a new dedicated PR for your changes, @refnil?
In this way, contributors can review the changes separately while I can cherry-pick them until it is merged to master.

**Comment by vale981** [comment]
I am a nix-noob, but would like to help to test this PR.
I would locally apply all associated PRs and then try to build the package. Is this how it's done, or is there a more efficient way?

**Comment by omasanori** [comment]
Thank you for your interest, @vale981!
Ideally, each PR is independent and can be applied to master alone. This one had not been such a PR, but remaining one was merged some hours ago. Thus, you can try this PR by `nix-shell -p nixpkgs-review --run "nixpkgs-review pr 101447"` (unless you are using `nixUnstable` but you can refer to `nix shell --help` then :wink:).

In general, `nix-shell -p nixpkgs-review --run "nixpkgs-review pr <the ID of PR>"` applies the PR on top of the current `master` branch pushed at `NixOS/nixpkgs` and builds everything it will affect: the packages, their dependencies if there are no cache files on `cache.nixos.org` and their reverse-dependencies (the packages depending on them, directly or indirectly). If it succeeds, you can test the resulting executables and so on.

See https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions for more description.

**Comment by omasanori** [comment]
So, thanks to helps from all of you, now `sage-tests` should *run*... with many failures.
They look not trivial (on my machine, the first one was an unhandled SIGFPE from an ECL-compiled library) but let's tackle with them.

**Comment by vale981** [comment]
I am getting:
```
builder for '/nix/store/c3i0kp4fazvi1d783f4pd2szwf0bfs9z-sagedoc-9.2.drv' failed with exit code 1; last 10 log lines:
      import sphinx.util.console
    File "/nix/store/s5x3vm0nshc74crkywlsbrxk161w7lm1-python3.8-sphinx-3.0.3/lib/python3.8/site-packages/sphinx/__init__.py", line 20, in <module>
      from .deprecation import RemovedInNextVersionWarning
    File "/nix/store/s5x3vm0nshc74crkywlsbrxk161w7lm1-python3.8-sphinx-3.0.3/lib/python3.8/site-packages/sphinx/deprecation.py", line 14, in <module>
      from typing import Any, Dict
    File "/nix/store/f57idsla85sxxgl9r5rfdzlgmqz4m0xj-python3.8-typing-3.7.4.3/lib/python3.8/site-packages/typing.py", line 1359, in <module>
      class Callable(extra=collections_abc.Callable, metaclass=CallableMeta):
    File "/nix/store/f57idsla85sxxgl9r5rfdzlgmqz4m0xj-python3.8-typing-3.7.4.3/lib/python3.8/site-packages/typing.py", line 1007, in __new__
      self._abc_registry = extra._abc_registry
  AttributeError: type object 'Callable' has no attribute '_abc_registry'
```

Looks like an incompatible python version.

**Comment by omasanori** [comment]
@vale981 Thank you for testing!
The error in Sphinx seems well-known one with the `typing` package triggered and newer Python where there is built-in version of the library: https://github.com/python/typing/issues/573. I will check if it is from `sagedoc` or `sphinx`.
Regarding `sage-tests`, it eats resources so much and takes hours so it is difficult to say it is a bug or usual run. My machine gets stuck at a different test...

**Comment by vale981** [comment]
Is it normal that there isn't any cpu activity for long stretches of time while running the tests?

**Comment by omasanori** [comment]
I guess it is not, but I am not sure as I am also a new member trying to fix the package with help from veterans.

Any thoughts? @timokau

**Comment by timokau** [comment]
> Regarding sage-tests, it eats resources so much and takes hours so it is difficult to say it is a bug or usual run. My machine gets stuck at a different test...

It might help to [run the test in isolation](https://github.com/NixOS/nixpkgs/pull/70474#issuecomment-714360025).

> Is it normal that there isn't any cpu activity for long stretches of time while running the tests?

No, at least I never noticed that. And I think I would have. If I had to guess, I'd say sage is waiting for some external process that isn't giving the expected reply. It uses `pexpect` to communicate with some dependencies, which can be brittle since it can depend on the exact formatting of the dependency's output. 

You can re-enable timeouts in the test-suite in `sage-tests.nix`. It currently disables them by passing `--timeout=0` to ensure deterministic/reproducible runs on slow or "overworked" machines. If you enable timeouts it should tell you which test is hanging.

**Comment by collares** [comment]
Using the above idea, I ran 
```nix-build -E 'with (import ./. {}); sage.tests.override { files = [ "src/doc/en/a_tour_of_sage/index.rst" ]; }```

and got SIGFPEs whose backtraces end with
```
/nix/store/iv42zh7skbps523bfy4h56nbygxqygi5-python3-3.8.6-env/lib/python3.8/site-packages/cysignals/signals.cpython-38-x86_64-linux-gnu.so(+0x8cbb)[0x7ffff5b1dcbb]
/nix/store/iv42zh7skbps523bfy4h56nbygxqygi5-python3-3.8.6-env/lib/python3.8/site-packages/cysignals/signals.cpython-38-x86_64-linux-gnu.so(+0x8d69)[0x7ffff5b1dd69]
/nix/store/iv42zh7skbps523bfy4h56nbygxqygi5-python3-3.8.6-env/lib/python3.8/site-packages/cysignals/signals.cpython-38-x86_64-linux-gnu.so(+0xcb2e)[0x7ffff5b21b2e]
/nix/store/a6rnjp15qgp8a699dlffqj94hzy1nldg-glibc-2.32/lib/libpthread.so.0(+0x12700)[0x7ffff7c2b700]
/nix/store/wba5k0csxlz9am2rzapwm44raq9qdc23-ecl-20.4.24/lib/libecl.so.20.4(ecl_extend_hashtable+0x11d)[0x7fff96c1ec9d]
```

Together with https://trac.sagemath.org/ticket/22191, this seems to suggest ecl needs to be updated.

**Comment by collares** [comment]
According to https://trac.sagemath.org/ticket/22191#comment:237, ECL needs to be patched to fix the SIGFPE crashes.

**Comment by collares** [comment]
With [this commit](https://github.com/collares/nixpkgs/commit/fda0fe9550e87a509df8e3d1e81cdd8e7d8d1af6) on top of the current PR, I get some "file not found" errors from the doctest harness and some failing tests but mostly the testsuite runs fine. I will update this comment with test results when it finishes.

**Comment by omasanori** [comment]
@timokau Indeed.
@collares Thank you for your investigation! I overlooked that ticket. I hope I will be able to check this or next week end.

(Unlike European culture, December in Japan does not mean holiday but *"work on last-minute tasks of the year, by the near-end of the year"* thing...)

**Comment by collares** [comment]
An update: https://github.com/collares/nixpkgs/tree/sage-update has only three failing test files at the moment: 

- `src/sage/interfaces/tests.py`: tests for ecl's behavior when both stderr and stdout get redirected to /dev/full. I tried to apply `write_error.patch` (see commit) but it didn't fix the edge case.
- `src/doc/en/thematic_tutorials/sandpile.rst` and `src/sage/sandpiles/sandpile.py`: for both files, the dict printing order is "wrong" (1, 2, 3, 0 instead of 0, 1, 2, 3) locally. I suspect some ordering problem causes the pretty printers for `SandpileConfig` and `SandpileDivisors` to not use the `_sorted_dict_pprinter_factory('{', '}')` pretty printer that was registered for dicts in `src/sage/doctest/forker.py`.

@omasanori If I fix the remaining failures before the weekend, is it OK for me to submit a new PR for review? You did most of the work and of course I would mention that in the PR. It would be completely understandable if you wanted to submit it yourself, so I am fine with waiting too.

**Comment by omasanori** [comment]
No problem, please go ahead open a new PR. I will be truly grad to test and review your changes.
Kudos to @collares and everyone involved in!

**Comment by omasanori** [comment]
Superseded by https://github.com/NixOS/nixpkgs/pull/105615

**Pull request closed without merging**